EventFilter/SiPixelRawToDigi/src/PixelDataFormatter.cc

```diff
@@ -193,22 +213,53 @@ void PixelDataFormatter::interpretRawData(bool& errorsInEvent, int fedId, const
     }
     if unlikely(skipROC) continue;
     
-    
-    int dcol = (ww >> DCOL_shift) & DCOL_mask;
-    int pxid = (ww >> PXID_shift) & PXID_mask;
+
     int adc  = (ww >> ADC_shift) & ADC_mask;
-    
-    LocalPixel::DcolPxid local = { dcol, pxid };
-    if unlikely(!local.valid()) {
-      LogDebug("PixelDataFormatter::interpretRawData") 
-	<< "status #3";
-      errorsInEvent = true;
-      errorcheck.conversionError(fedId, &converter, 3, ww, errors);
-      continue;
-    }
-    
-    GlobalPixel global = rocp->toGlobal( LocalPixel(local) );
+    LocalPixel *local=NULL;
+
+    if(phase1 && layer==1) { // special case for layer 1ROC
+      // for l1 roc use the roc column and row index instead of dcol and pixel index.
+      int col = (ww >> COL_shift) & COL_mask;
+      int row = (ww >> ROW_shift) & ROW_mask;
+      //if(DANEK) cout<<" layer 1: raw2digi "<<link<<" "<<roc<<" "
+      //  <<col<<" "<<row<<" "<<adc<<endl;      
+
+      LocalPixel::RocRowCol localCR = { row, col }; // build pixel
+      //if(DANEK)cout<<localCR.rocCol<<" "<<localCR.rocRow<<endl;
+      if unlikely(!localCR.valid()) {
+	  LogDebug("PixelDataFormatter::interpretRawData") 
+	    << "status #3";
+	  errorsInEvent = true;
+	  errorcheck.conversionError(fedId, &converter, 3, ww, errors);
+	  continue;
+	}
+      local = new LocalPixel(localCR); // local pixel coordinate 
+      //if(DANEK) cout<<local->dcol()<<" "<<local->pxid()<<" "<<local->rocCol()<<" "<<local->rocRow()<<endl;
+
+    } else { // phase0 and phase1 except bpix layer 1
+      int dcol = (ww >> DCOL_shift) & DCOL_mask;
+      int pxid = (ww >> PXID_shift) & PXID_mask;
+      //if(DANEK) cout<<" raw2digi "<<link<<" "<<roc<<" "
+      //<<dcol<<" "<<pxid<<" "<<adc<<" "<<layer<<endl;
+      
+      LocalPixel::DcolPxid localDP = { dcol, pxid };
+      //if(DANEK) cout<<localDP.dcol<<" "<<localDP.pxid<<endl;
+
+      if unlikely(!localDP.valid()) {
+	  LogDebug("PixelDataFormatter::interpretRawData") 
+	    << "status #3";
+	  errorsInEvent = true;
+	  errorcheck.conversionError(fedId, &converter, 3, ww, errors);
+	  continue;
+	}
+      local = new LocalPixel(localDP);
+      //if(DANEK) cout<<local->dcol()<<" "<<local->pxid()<<" "<<local->rocCol()<<" "<<local->rocRow()<<endl;
+    }    
+
+    //GlobalPixel global = rocp->toGlobal( LocalPixel(local) );
+    GlobalPixel global = rocp->toGlobal( *local ); // global pixel coordinate (in module)
```

Comment:
Carl,
Yes, indeed there is a problem. Thanks for spotting. I will fix it.
How does one use 
std::unique_ptr
is there a good example?
Danek

> On 7 Sep 2016, at 23:41, Carl Vuosalo <notifications@github.com> wrote:
> 
> In EventFilter/SiPixelRawToDigi/src/PixelDataFormatter.cc:
> 
> > +      LocalPixel::DcolPxid localDP = { dcol, pxid };
> > +      //if(DANEK) cout<<localDP.dcol<<" "<<localDP.pxid<<endl;
> > +
> > +      if unlikely(!localDP.valid()) {
> > +	  LogDebug("PixelDataFormatter::interpretRawData") 
> > +	    << "status #3";
> > +	  errorsInEvent = true;
> > +	  errorcheck.conversionError(fedId, &converter, 3, ww, errors);
> > +	  continue;
> > +	}
> > +      local = new LocalPixel(localDP);
> > +      //if(DANEK) cout<<local->dcol()<<" "<<local->pxid()<<" "<<local->rocCol()<<" "<<local->rocRow()<<endl;
> > +    }    
> > +
> > +    //GlobalPixel global = rocp->toGlobal( LocalPixel(local) );
> > +    GlobalPixel global = rocp->toGlobal( *local ); // global pixel coordinate (in module)
> 
> The static analyzer finds a path through this code that results in rocp being null at this point. A check for a null pointer or an assert(rocp) could be added here, or the logic could be changed to ensure rocp is always valid.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>